pkg/cloud/gcp/actuators/machine/reconciler.go

```diff
@@ -278,3 +294,68 @@ func isNotFoundError(err error) bool {
 	}
 	return false
 }
+
+func fmtInstanceSelfLink(project, zone, name string) string {
+	return fmt.Sprintf(instanceLinkFmt, project, zone, name)
+}
+
+func (r *Reconciler) instanceExistsInPool(instanceLink string, pool string) (bool, error) {
+	// Get target pool
+	tp, err := r.computeService.TargetPoolsGet(r.projectID, r.providerSpec.Region, pool)
+	if err != nil {
+		return false, fmt.Errorf("unable to get targetpool: %v", err)
+	}
+
+	// Check if instance is in target pool already
+	if tp.Instances != nil {
+		for _, link := range tp.Instances {
+			if instanceLink == link {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+type poolProcessor func(instanceLink, pool string) error
+
+func (r *Reconciler) processTargetPools(desired bool, poolFunc poolProcessor) error {
+	if len(r.providerSpec.TargetPools) == 0 {
```

Comment:
You can drop this condition; it is safe to iterate over a nil slice.

Reply:
That's true, but saving the call to generate instanceSelfLink is string manipulation and is expensive compared to this check.

Reply:
Move that to where the condition holds true.

Reply:
> That's true, but saving the call to generate instanceSelfLink is string manipulation and is expensive compared to this check.

OK; but given the amount of general logging we do I think calling fmt here for the nil cases... is no worse than just another log statement.